### DISCUSSION TITLE
Improve loading time of multiple models

### DISCUSSION THREAD

**Discussion by eladbitton** [explosion/spaCy#9067] [Help: Best practices]
I am using spacy to split sentences using the 'sentencizer' model.

I am loading a model for any of the following languages:
"zh", "da", "nl", "en", "fr", "de", "it", "pt", "ru", "es", "xx"

Loading these models one by one took on my pc about 55 seconds.
Making it parallel using a thread pool took about 51 seconds.

My code:
```

import spacy
from concurrent.futures import ThreadPoolExecutor

MULTILINGUAL_MODEL_NAME = "xx"
SUPPORTED_LANGUAGES = set(["zh",
                           "da",
                           "nl",
                           "en",
                           "fr",
                           "de",
                           "it",
                           "pt",
                           "ru",
                           "es",
                           MULTILINGUAL_MODEL_NAME])  # Multilingual


def load_model(lang):
    model = spacy.util.get_lang_class(lang)(tagger=False, parser=False, entity=False, exclude=["tok2vec", "tagger", "parser", "attribute_ruler", "lemmatizer"])
    model.add_pipe('sentencizer')
    return model


def load_models():
    models = dict()
    with ThreadPoolExecutor(15) as tp:
        futures = []
        for lang in SUPPORTED_LANGUAGES:
            f = tp.submit(load_model, lang)
            futures.append((lang, f))

        for lang, future in futures:
            models[lang] = future.result()

    return models
```

Is there any way to make it load faster?

**Comment by polm** [comment]
You code took 15s on my machine, but if I change this line:

    model = spacy.util.get_lang_class(lang)(tagger=False, parser=False, entity=False, exclude=["tok2vec", "tagger", "parser", "attribute_ruler", "lemmatizer"])

to be like this:

    model = spacy.blank(lang)

It took 5s instead. So maybe try that?

**Comment by eladbitton** [comment]
@polm thanks for the advice, I am now using blank instead.

Interestingly, I've added a print to show how much time each model took to load:
```
def load_model(lang):
    # model = spacy.util.get_lang_class(lang)(tagger=False, parser=False, entity=False, exclude=["tok2vec", "tagger", "parser", "attribute_ruler", "lemmatizer"])
    a = time()
    model = spacy.blank(lang)
    model.add_pipe('sentencizer')
    print(f"{lang} took: {time() - a}")
    return model
```

And these are the results:
xx took: 1.2940025329589844
zh took: 1.4933297634124756
ru took: 2.789743423461914
it took: 4.069791555404663
de took: 4.144196033477783
pt took: 4.674798250198364
en took: 5.750870704650879
da took: 6.3298139572143555
es took: 6.509328365325928
nl took: 8.617478609085083
fr took: 42.575430393218994

Somehow the French model is very slow.

This is what i have installed:
fr-core-news-sm @ https://github.com/explosion/spacy-models/releases/download/fr_core_news_sm-3.1.0/fr_core_news_sm-3.1.0-py3-none-any.whl

Any idea why is that?

**Reply by polm** [reply]
I am surprised the French model is that slow, but it looks like most of the loading time is in dealing with the tokenizer exceptions, which are more complicated than many other languages. I kind of assumed some of the mdoels with external dependencies, like Chinese, would be slower.

I will say that it's a lot faster on my machine, and only takes about 7s.

Still, since that's pretty weird we'll take a look at it, thanks for the details!
